### PR TITLE
Allow major Etcd upgrades with safe roll-back

### PR DESCRIPTION
**Pull request by davidmccormick**
Safer major etcd upgrades by spinning up new etcds and performing a migration (copy of all kubernetes data)

Similar to the approach used during the stack migration but this time the major/minor version of etcd is used to control migration e.g 3.2.x -> 3.3.x will cause a migration and the migration is handled within the same etcd stack.  It is safer because should the CF roll fail the previous etcd's should still be available to roll-back to.  It works by embedding the version into the etcd cloudformation logical names so that version upgrades generates a new set of etcds, e.g.

```
+00:00:58	CREATE_IN_PROGRESS      		Etcdv3dot3i2ENI
+00:00:58	CREATE_IN_PROGRESS      		Etcdv3dot3i2LC
+00:00:58	CREATE_IN_PROGRESS      		Etcdv3dot3i0LC
+00:00:58	CREATE_IN_PROGRESS      		Etcdv3dot3i1LC
+00:00:59	CREATE_IN_PROGRESS      		Etcdv3dot3i2ENI       	"Resource creation Initiated"
+00:00:59	CREATE_IN_PROGRESS      		Etcdv3dot3i1ENI
+00:00:59	CREATE_IN_PROGRESS      		Etcdv3dot3i2LC        	"Resource creation Initiated"
+00:00:59	CREATE_IN_PROGRESS      		Etcdv3dot3i0LC        	"Resource creation Initiated"
+00:00:59	CREATE_IN_PROGRESS      		Etcdv3dot3i1LC        	"Resource creation Initiated"
+00:00:59	CREATE_IN_PROGRESS      		Etcdv3dot3i0ENI
+00:00:59	UPDATE_IN_PROGRESS      		IAMRoleEtcd
+00:00:59	CREATE_IN_PROGRESS      		Etcdv3dot3i1ENI       	"Resource creation Initiated"
+00:00:59	CREATE_COMPLETE         		Etcdv3dot3i2LC
+00:00:59	CREATE_COMPLETE         		Etcdv3dot3i0LC
+00:00:59	CREATE_COMPLETE         		Etcdv3dot3i1LC
+00:00:59	CREATE_IN_PROGRESS      		Etcdv3dot3i0ENI       	"Resource creation Initiated"
+00:01:09	UPDATE_COMPLETE         		IAMRoleEtcd
+00:01:15	CREATE_COMPLETE         		Etcdv3dot3i2ENI
+00:01:15	CREATE_COMPLETE         		Etcdv3dot3i1ENI
+00:01:15	CREATE_COMPLETE         		Etcdv3dot3i0ENI
+00:01:17	CREATE_IN_PROGRESS      		Etcdv3dot3i2EBS
+00:01:17	CREATE_IN_PROGRESS      		Etcdv3dot3i2EBS       	"Resource creation Initiated"
+00:01:17	CREATE_IN_PROGRESS      		Etcdv3dot3i1EBS
+00:01:18	CREATE_IN_PROGRESS      		Etcdv3dot3i1EBS       	"Resource creation Initiated"
+00:01:18	CREATE_IN_PROGRESS      		Etcdv3dot3i0EBS
+00:01:18	CREATE_IN_PROGRESS      		Etcdv3dot3i0EBS       	"Resource creation Initiated"
+00:01:33	CREATE_COMPLETE         		Etcdv3dot3i2EBS
+00:01:34	CREATE_COMPLETE         		Etcdv3dot3i1EBS
+00:01:34	CREATE_COMPLETE         		Etcdv3dot3i0EBS
+00:01:35	CREATE_IN_PROGRESS      		Etcdv3dot3i2
+00:01:36	CREATE_IN_PROGRESS      		Etcdv3dot3i1
+00:01:36	CREATE_IN_PROGRESS      		Etcdv3dot3i2          	"Resource creation Initiated"
+00:01:36	CREATE_IN_PROGRESS      		Etcdv3dot3i0
+00:01:36	CREATE_IN_PROGRESS      		Etcdv3dot3i1          	"Resource creation Initiated"
+00:01:37	CREATE_IN_PROGRESS      		Etcdv3dot3i0          	"Resource creation Initiated"
+00:03:59	CREATE_IN_PROGRESS      		Etcdv3dot3i2          	"Received SUCCESS signal with UniqueId i-0797f12fe4b3d5840"
+00:04:00	CREATE_COMPLETE         		Etcdv3dot3i2
+00:04:00	CREATE_IN_PROGRESS      		Etcdv3dot3i0          	"Received SUCCESS signal with UniqueId i-09caf63ec80dd3f37"
+00:04:01	CREATE_COMPLETE         		Etcdv3dot3i0
+00:04:39	CREATE_IN_PROGRESS      		Etcdv3dot3i1          	"Received SUCCESS signal with UniqueId i-02b447cadb74b4265"
+00:04:40	CREATE_COMPLETE         		Etcdv3dot3i1
```
kube-aws uses a new tag on the etcds ` kube-aws:etcd_upgrade_group` to look for etcds in the same upgrade group as the requested deployment; if the cluster exists (i.e. cluster etcd stack exists) but there are no matching etcd_upgrade_group instances then kube-aws will trigger the migration mode.  In migration mode the leader of the new etcd cluster exports the kubernetes registry from the existing/old etcds and restores them into the new cluster.

Features
* Brings all new etcds up at same time during a migration.
* When an etcd has an attached NIC use that address rather than the machine's private dnsname
* Update Etcd to 3.3.17 release
* Fix etcdadm so that it can still detect cluster healthy now written to stderr
* Update etcd migration to respect keys with leases
* Fix building etcd endpoints where the interfaces are listed in different orders
* Move to a two export process for retrieving keys and values using etcdctl 'json'  export type for key/value data, and then again using its 'fields' export type in order to successfully extract key/lease data.  * Process the two files back together with a nod to performance.

### CODE REVIEW

File: pkg/model/context.go [hunk 1]
```diff
@@ -62,19 +63,69 @@ func (s *Context) InspectEtcdExistingState(c *Config) (api.EtcdExistingState, er
...
+	if state.StackExists {
+		if state.EtcdMigrationEnabled, err = s.isAMajorEtcdUpgrade(c); err != nil {
+			return state, fmt.Errorf("failed to check existing etcd major minor version: %v", err)
 		}
-		if state.EtcdMigrationExistingEndpoints != "" {
-			state.EtcdMigrationEnabled = true
+		if state.EtcdMigrationEnabled {
+			logger.Warn("Performing a Major Etcd Version Upgrade: -")
+			logger.Warn("To do this we will spin up new etcd servers and then export the existing kubernetes state to them.")
+			logger.Warn("There will be cluster disruption until all of your existing controllers have rolled.")
```
Comment (dominicgunn) [inline pkg/model/context.go]: What is the window of disruption here? Are we talking service outages?

File: pkg/model/context.go [hunk 2]
```diff
@@ -62,19 +63,69 @@ func (s *Context) InspectEtcdExistingState(c *Config) (api.EtcdExistingState, er
...
+			{
+				Name:   aws.String("tag:kube-aws:role"),
+				Values: []*string{aws.String("etcd")},
+			},
+			{
+				Name:   aws.String(clusterTag),
+				Values: []*string{aws.String("owned")},
+			},
+			{
+				Name:   aws.String("tag:kube-aws:etcd_upgrade_group"),
```
Comment (dominicgunn) [inline pkg/model/context.go]: Is this backward compatible with older clusters? I guess it'd just return 0 instances right?

File: pkg/model/context.go [hunk 3]
```diff
@@ -62,19 +63,69 @@ func (s *Context) InspectEtcdExistingState(c *Config) (api.EtcdExistingState, er
...
+	if err != nil {
+		return false, fmt.Errorf("can't lookup ec2 instances: %v", err)
+	}
+	if resp == nil {
+		return false, nil
+	}
+
+	logger.Debugf("<- received %d instances from AWS", len(resp.Reservations))
+	if len(resp.Reservations) == 0 {
+		logger.Debugf("There are 0 instances matching major-minor version %s - this is an upgrade...", c.Etcd.Cluster.MajorMinorVersion())
```
Comment (dominicgunn) [inline pkg/model/context.go]: Looks like this catches the backwards compatibility issues, cool.

File: pkg/model/context.go [hunk 4]
```diff
@@ -62,19 +63,69 @@ func (s *Context) InspectEtcdExistingState(c *Config) (api.EtcdExistingState, er
...
-		if state.EtcdMigrationExistingEndpoints != "" {
-			state.EtcdMigrationEnabled = true
+		if state.EtcdMigrationEnabled {
+			logger.Warn("Performing a Major Etcd Version Upgrade: -")
+			logger.Warn("To do this we will spin up new etcd servers and then export the existing kubernetes state to them.")
+			logger.Warn("There will be cluster disruption until all of your existing controllers have rolled.")
+			logger.Warn("If the cloudformation update fails (at any point) then we will roll back to the original etcd servers.")
+			logger.Warn("You MAY lose/rollback changes that are made to the cluster AFTER the etcd export has been performed!")
+			logger.Warn("This operation is best scheduled for a quiet time or in an outage window.")
+			if state.EtcdMigrationExistingEndpoints, err = s.lookupExistingEtcdEndpoints(c); err != nil {
```
Comment (dominicgunn) [inline pkg/model/context.go]: If we're going to bomb out on `lookupExistingEtcdEndpoints`, should we try and do that before telling the user that we're `Performing a Major Etcd Upgrade`? 

As a layman, I might be concerned that something may have changed before the thing blew up.

File: pkg/model/context.go (hunk 1 above)
Comment (davidmccormick) [reply pkg/model/context.go]: Good question - we are talking about a service outage of the kubernetes api of several minutes (good case) longer if a roll-back is triggered after api-servers have been rolled.

File: pkg/model/context.go (hunk 2 above)
Comment (davidmccormick) [reply pkg/model/context.go]: Yes, this change would include the first migratioin - to 3.3.17

File: pkg/model/context.go (hunk 4 above)
Comment (davidmccormick) [reply pkg/model/context.go]: Good idea - I've swapped it around so we announce the migration only if the lookup is successful.

**Pull request merged**